### ISSUE TITLE
Documenting material entities

### ISSUE THREAD

**Issue by SilverfishVR** [kasenvr/vircadia-docs-sphinx#22]
Now that the Procedural material documentation got some love it is painfully obvious that the "normal" material is lacking and wrong in some places.

Initially I would like to make an entry in the docs that describe how material entities work, including ALL valid properties and how they interact, so I am inviting anyone to review/correct my findings or point me to additional documentation.

Materials on imported models adds an extra layer of logic as to how different properties are mapped that I honestly don't fully understand and currently it's a bit of a limbo where GLTF support is still kinda iffy, vanilla blender fbx export has issues, and the metaverse plugin fbx export is currently broken. That is why I, for now will only try to document the material entity properties, since that is in the end, what all imported entities will be mapped to.

In the api docs all I find is this: https://apidocs.projectathena.dev/Graphics.html#.Material it is somewhat unhelpful and weirdly missing some properties like albedo and emissive, but listing some I don't know how to use like texCoordTransform and materialParams.
I believe the api docs are generated from comments in source so I would need help updating this, I used Doxygen way back when, but this feels different.

In the docs there is the PBR mterials guide: https://docs.projectathena.dev/create/3d-models/pbr-materials-guide.html wich is also lacking and outdated in places.

These are the material data properties for material entities I have spend way too much time testing:

Property | Type | Status
----------|-------| -------|
"name" | [string] | OK.
"model" |  [string]  | OK. Valid values: "hifi_pbr", "hifi_shader_simple" others? 
"opacity" | [0,1f] | OK.
"roughness" |  [0,1f] | works, but only if metallic is > 0 or scattering is enabled (map or value) or emissive is enabled (map or value). overridden by "roughnessMap" or "glossMap".
"metallic" | [0,1f] | OK.
"scattering" | [0,1f] | OK. I think, it does do something.
"unlit" |  [true, false] | OK. must be "true" or "false" not 0/1
"emissive" | [?] | OK. range? 0-1 looks like normal emissive but 1-128 seems to amplify bloom (if enabled in zone)?
"emissiveMap" | [texture] | OK.
"albedoMap": | [texture] | OK.
"opacityMap" | [texture] | the texture itself is not applied, see below. 
"opacityMapMode" | [string] | OK. valid values: OPACITY_MAP_MASK, OPACITY_MAP_BLEND, OPACITY_MAP_OPAQUE
"opacityCutoff" | [0,1f] | OK. non-linear.
"metallicMap" | [texture] | OK. overrides specularMap
"specularMap" | [texture] | OK. maps to metallicMap if one is not defined
"roughnessMap" | [texture] | OK.
"glossMap" | [texture] | OK. maps to (inverted)roughnessMap if roughnessMap is not defined.
"normalMap" | [texture] |OK.
"bumpMap" | [texture] | OK.
"occlusionMap" | [texture] | OK I think. Does map to occlusionMap, really subtle effect not quite shure.
"lightMap" |  [texture] | OK I think. Does map to lightMap, seems to work.
"scatteringMap" | [texture] | OK, does not override scattering value.
"cullFaceMode" | [string] |  OK. Valid values: CULL_NONE, CULL_FRONT, CULL_BACK
"defaultFallthrough" | [true/false] | semi OK. see below(must be "true" or "false" not 0/1)
"procedural" | [string] | OK. all the custom shader stuff, out of scope.

These I do not know how to test:

Property | Type | Status
----------|-------| -------|
"texCoordTransform0" | ? | 
"texCoordTransform1" | ? | 
"lightmapParams" | ? | 
"materialParams" | ? | 

And then there is the **"materialVersion"** it is not in the materials array, it is always 1, and seems to do nothing?

## About fallthrough:
I think all properties, exept for "name", "model", "defaultFallthrough" and "procedural" are supposed to be able set to "fallthrough", mostly works but it does not work on roughness.
I presume that the reason that the api docs lists "string" as a secondary property type is because they can also be "fallthrough"?

DefaultFallthrough is rather buggy, or maybe I just don't understand how it is supposed to work but I don't think so, albebo, for example, works as I expect.
setting defaultFallthrough to true on the highest priority material shows few bugs:
1. if opacity is 1, it will fallthrough, any value < 1 works.
2. roughness and metallic seems to not work, goes to default value.
3. emissive works but reading back material properties reflects default values.
4. if unlit is false it will always fallthrough.
And probably many more, the actual property value is 

## About Opacity:
Using a masked or alpha texture, opacity does not show unless:
1.  opacityMap is set to the same texture as albeboMap, or
2.  opacity is less than 1, or
3.  *opacityMapMode* is set to either OPACITY_MAP_MASK or OPACITY_MAP_BLEND.

## TL:DR
I will document how material entities work if you help me with the unknowns, then we build from there

**Comment by ctrlaltdavid** [comment]
Note: There are some additions to material support in HiFi master which aren't yet included in the current 0.86.0-based Athena.
- Material Entities can specify face culling method, support GLTF doubleSided property 
- Procedural Material Entities can specify a "vertexShaderURL"
These may or may not be included in the HiFi master JSDoc.

**Comment by vegaslon** [comment]
There is also this https://apidocs.vircadia.dev/Entities.html#.EntityProperties-Material

**Comment by vegaslon** [comment]
materialVersion is for the json materials which are cached when they are loaded. that allows you to tweak the materials and not run into the cached version.

**Comment by vegaslon** [comment]
"texCoordTransform0" and 1 are the materialmappings set by the material entity properties.
emissive is a vector and the range from 0-1 is normal range of light reflection which is what makes this PBR. Setting above 1 is part of the bloom, which is not technically PBR but still a cool feature. Emissive above 1 basically is used to boost bloom based on how far it is set above the bloom threshold on the zone, which is how much light to start applying bloom to. Setting a bloom threshold above 1 in most game engines means that in a normal PBR sense there will be no bloom rendered unless you set the emissive for a entity above 1 as well.

**Comment by SilverfishVR** [comment]
https://apidocs.projectathena.dev/Graphics.html#.Material  was updated so it's fine I guess
